README
Installation
------------

To install libfap on a linux system use the normal procedure:

./configure [--prefix=/opt or other options]
make
<become root if necessary>
make install

For detailed installation instructions see the INSTALL file.

Usage
-----

Usage info is available at http://pakettiradio.net/libfap/docs/ . A local
copy of the docs can be generated by running doxygen with the attached
Doxyfile. When in doubt, see the fap.h header file directly.

Differences
-----------

There is one intentional difference between the original Perl module and
libfap. The Bob has apparently said that a space or a / can be placed in the
beginning of station commment in order to separate it clearly from APRS
data. FAP.pm removes these characters and also other whitespaces from
beginning and end of the comment. Libfap does not remove these, as I believe
that making things look nice is a job for user of the parser, not the parser
itself. This way you can also use the original comment as-is, if you want
to.

There is also one "accidental" difference. The error message field of the
hash in Perl module is used only once in one specific NMEA error case. To
avoid confusion with error code field and messages corresponding the codes,
the error message field has been dropped from libfap.


73 de OH2GVE


ChangeLog
libfap 1.4

   * Fixed crash in parse_wx_peet_logging
   * Fixed timestamps not being returned in UTC
   * Added raw_timestamp field, which contains original time string
   * Improved telemetry handling

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Sun, 23 Nov 2014 13:38:09 +0200

libfap 1.3

   * Fixed RAD2DEG() used in distance calculation
   * Fixed setting of has-been-repeated bit in tnc2_to_ax25()

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Sun, 20 Nov 2011 08:12:37 +0200

libfap 1.2

   * Added metadata file for pkg-config

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Sun, 22 May 2011 19:34:17 +0300

libfap 1.1

   * Changed code-to-text explainging functions returning char* to use
     pre-allocated buffer instead
   * Removed rarely used error_message field from fap_packet_t
   * Outdated smoke test files removed

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Sat, 26 Feb 2011 11:28:47 +0200

libfap 1.0

   * Re-implemented WX report software ID detection to match Perl module
   * Fixed position resolution calculation for whole degree ambiguity case
   * Added error message for too short compressed packet
   * Clarified licensing terms

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Mon, 17 Jan 2011 14:01:15 +0200

libfap 0.9

   * Now matching Perl module version 1.17
     * IPv6 support added
     * Comment cleanup NOT implemented, see README for more info

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Tue, 12 Oct 2010 08:08:29 +0300

libfap 0.8

   * Implemented improvements introduced in Perl module version 1.15

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Thu, 06 May 2010 12:36:29 +0300

libfap 0.7

   * Empty string turned out to be invalid regex on POSIX. Some regexes were
     adjusted to comply with this.
   * Added functions for converting between raw AX.25-frames and TNC-2
     compatible UI-frames
   * Implemented automated tests using XS language and tests of the original
     Perl module. And thanks to this:
   * Fixed position resolution calculation for normalpos and mic-e packets
     (it didn't happen at all)
   * Fixed position calculation for normalpos packets with ambiguity of 3
   * Normal or compressed pos is now detected also from first 40 chars of body
   * Changed type of speed value from unsigned int to double
   * Changed type of altitude value from int to double
   * Fixed wx parser to support reports which begin with 'g'
   * Changed mic-e packet symtable and symcode fields (they were mixed)
   * Fixed mic-e packet speed calculation for large speeds
   * Changed timestamp of GPRMC packet from local time to UTC
   * Changed lat and lon of GPRMC packet with each other (they were mixed)
   * Added support for snow and softare type into wx report
   * Fixed rounding errors from Peet type wx packet parsing
   * Improved message id detection to handle messages with non-printable chars
   * Added error code for non-APRS packet
   * Fixed incorrectly applied Base-91-DAO

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Sat, 03 Apr 2010 10:24:18 +0300

libfap 0.6

   * Implemented improvements introduced in new Perl module version (1.13)

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Fri, 11 Dec 2009 20:13:27 +0200

libfap 0.5

   * Type incompatibility warning fixed from parse_mice()

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Tue, 20 Oct 2009 10:19:50 +0300

libfap 0.4

   * Memory leak and use of uninitialized variables fixed from tnc2_to_kiss

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Mon, 19 Oct 2009 09:58:05 +0300

libfap 0.3

   * Path validation works now correctly also when is_ax25 flag is set

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Sat, 17 Oct 2009 09:42:58 +0300

libfap 0.2

   * Adjusted licensing to Artistic license or GPL
   * Added checks for out-of-memory situation

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Thu, 08 Oct 2009 07:38:12 +0300

libfap 0.1

   * Initial release

 -- Tapio Aaltonen, OH2GVE <oh2gve@pakettiradio.net> Tue, 06 Oct 2009 18:16:27 +0300


src/fap.h
/* $Id: fap.h 187 2011-02-16 21:58:11Z oh2gve $
 *
 * Copyright 2005, 2006, 2007, 2008, 2009, 2010 Tapio Sokura
 * Copyright 2007, 2008, 2009, 2010 Heikki Hannikainen
 *
 * Perl-to-C modifications 
 * Copyright 2009, 2010, 2011 Tapio Aaltonen
 *
 * This file is part of libfap.
 *
 * Libfap is free software; you can redistribute it and/or modify it under the
 * terms of either:
 *
 * a) the GNU General Public License as published by the Free Software
 * Foundation; either version 1, or (at your option) any later
 * version, or
 * 
 * b) the "Artistic License". 
 * 
 * Both licenses can be found in the licenses directory of this source code
 * package.
 *
 * APRS is a registered trademark of APRS Software and Bob Bruninga, WB4APR.
*/

/**
 * \file fap.h
 * \brief Declaration of libfap's public interface.
 *
 * This file declares data types used to represent a single APRS packet.
 * Functions include the parser, some utility functions and library
 * initialization stuff.
 *
 * Note: it is very important to call fap_init() before calling any other
 * function. Also note that fap_cleanup() should be called when shutting
 * down, or memory leaks will occur.
 *
 * \author Tapio Aaltonen
*/

#ifndef FAP_H
#define FAP_H


#include <time.h>


#ifdef __cplusplus
extern "C" {
#endif 



/// Packet error code type.
typedef enum
{
	fapPACKET_NO,
	fapPACKET_SHORT,
	fapPACKET_NOBODY,
	
	fapSRCCALL_NOAX25,
	fapSRCCALL_BADCHARS,
	
	fapDSTPATH_TOOMANY,
	fapDSTCALL_NONE,
	fapDSTCALL_NOAX25,
	
	fapDIGICALL_NOAX25,
	fapDIGICALL_BADCHARS,
	
	fapTIMESTAMP_INV_LOC,
	fapTIMESTAMP_INV_OBJ,
	fapTIMESTAMP_INV_STA,
	fapTIMESTAMP_INV_GPGGA,
	fapTIMESTAMP_INV_GPGLL,
	
	fapPACKET_INVALID,
	
	fapNMEA_INV_CVAL,
	fapNMEA_LARGE_EW,
	fapNMEA_LARGE_NS,
	fapNMEA_INV_SIGN,
	fapNMEA_INV_CKSUM,
	
	fapGPRMC_FEWFIELDS,
	fapGPRMC_NOFIX,
	fapGPRMC_INV_TIME,
	fapGPRMC_INV_DATE,
	fapGPRMC_DATE_OUT,
	
	fapGPGGA_FEWFIELDS,
	fapGPGGA_NOFIX,
	
	fapGPGLL_FEWFIELDS,
	fapGPGLL_NOFIX,

	fapNMEA_UNSUPP,
	
	fapOBJ_SHORT,
	fapOBJ_INV,
	fapOBJ_DEC_ERR,
	
	fapITEM_SHORT,
	fapITEM_INV,
	fapITEM_DEC_ERR,
	
	fapLOC_SHORT,
	fapLOC_INV,
	fapLOC_LARGE,
	fapLOC_AMB_INV,
	
	fapMICE_SHORT,
	fapMICE_INV,
	fapMICE_INV_INFO,
	fapMICE_AMB_LARGE,
	fapMICE_AMB_INV,
	fapMICE_AMB_ODD,
	
	fapCOMP_INV,
	fapCOMP_SHORT,
	
	fapMSG_INV,
	
	fapWX_UNSUPP,
	fapUSER_UNSUPP,
	
	fapDX_INV_SRC,
	fapDX_INF_FREQ,
	fapDX_NO_DX,
	
	fapTLM_INV,
	fapTLM_LARGE,
	fapTLM_UNSUPP,
	
	fapEXP_UNSUPP,
	fapSYM_INV_TABLE,
	
	fapNOT_IMPLEMENTED,
	fapNMEA_NOFIELDS,
	
	fapNO_APRS
} fap_error_code_t;


/// Packet type type.
typedef enum
{
	fapLOCATION,
	fapOBJECT,
	fapITEM,
	fapMICE,
	fapNMEA,

	fapWX,
	fapMESSAGE,
	fapCAPABILITIES,
	fapSTATUS,
	fapTELEMETRY,
	fapTELEMETRY_MESSAGE,
	fapDX_SPOT,

	fapEXPERIMENTAL
} fap_packet_type_t;


/// Position format type.
typedef enum
{
	fapPOS_COMPRESSED,
	fapPOS_UNCOMPRESSED,
	fapPOS_MICE,
	fapPOS_NMEA
} fap_pos_format_t;


/// Weather report type.
typedef struct
{
	/// Wind gust in m/s.
	double* wind_gust;
	/// Wind direction in degrees.
	unsigned int* wind_dir;
	/// Wind speed in m/s.
	double* wind_speed;
	
	/// Temperature in degrees Celcius.
	double* temp;
	/// Indoor temperature in degrees Celcius.
	double* temp_in;
	
	/// Rain from last 1 hour, in millimeters.
	double* rain_1h;
	/// Rain from last day, in millimeters.
	double* rain_24h;
	/// Rain since midnight, in millimeters.
	double* rain_midnight;
	
	/// Relative humidity percentage.
	unsigned int* humidity;
	/// Relative inside humidity percentage.
	unsigned int* humidity_in;

	/// Air pressure in millibars.
	double* pressure;
	/// Luminosity in watts per square meter.
	unsigned int* luminosity;
	
	/// Show depth increasement from last day, in millimeters.
	double* snow_24h;
	
	/// Software type indicator.
	char* soft;
} fap_wx_report_t;



/// Telemetry report type.
typedef struct
{
	/// Id of report.
	unsigned int seq;
	/// First value.
	double val1;
	/// Second value.
	double val2;
	/// Third value.
	double val3;
	/// Fourth value.
	double val4;
	/// Fifth value.
	double val5;
	
	/// Telemetry bits as ASCII 0s and 1s. Unknowns are marked with question marks.
	char bits[8];
} fap_telemetry_t;



/// APRS packet type.
typedef struct
{
	/// Error code.
	fap_error_code_t* error_code;
	/// Packet type.
	fap_packet_type_t* type;
	
	/// Exact copy of the original packet, if such thing was given.
	char* orig_packet;
	/// Length of orig_packet.
	unsigned int orig_packet_len;
	
	/// Header part of the packet.
	char* header;
	/// Body of the packet. No null termination.
	char* body;
	/// Lenght of body.
	unsigned int body_len;
	/// AX.25-level source callsign.
	char* src_callsign;
	/// AX.25-level destination callsign.
	char* dst_callsign;
	/// Array of path elements.
	char** path;
	/// Number of path elements.
	unsigned int path_len;
	
	/// Latitude, west is negative.
	double* latitude;
	/// Longitude, south is negative.
	double* longitude;
	/// Position format.
	fap_pos_format_t* format;
	/// Position resolution in meters.
	double* pos_resolution;
	/// Position ambiguity, number of digits.
	unsigned int* pos_ambiguity;
	/// Datum character from !DAO! extension. 0x00 = undef.
	char dao_datum_byte;

	/// Altitude in meters.
	double* altitude;
	/// Course in degrees, zero is unknown and 360 is north.
	unsigned int* course;
	/// Land speed in km/h.
	double* speed;

	/// Symbol table designator. 0x00 = undef.
	char symbol_table;
	/// Slot of symbol table 0x00 = undef.
	char symbol_code;
	
	/// Zero for no messaging capability, 1 for yes.
	short* messaging;
	/// Destination of message.
	char* destination;
	/// The actual message text.
	char* message;
	/// Id of the message which is acked with this packet.
	char* message_ack;
	/// Id of the message which was rejected with this packet.
	char* message_nack;
	/// Id of this message.
	char* message_id;
	/// Station, object or item comment. No null termination.
	/**
	 * Here's a difference between Perl module and C library. Perl
	 * module removes whitespace characters from beginning and end of
	 * the comment.  C library returns comment exactly as seen in
	 * original packet.
	*/
	char* comment;
	/// Length of comment.
	unsigned int comment_len;

	/// Name of object or item in packet.
	char* object_or_item_name;
	/// Object or item status. 1 = alive, 0 = killed.
	short* alive;

	/// Zero if GPS has no fix, one if it has.
	short* gps_fix_status;
	/// Radio range of the station in km.
	unsigned int* radio_range;
	/// TX power, antenna height, antenna gain and possibly beacon rate.
	char* phg;
	/// Timestamp of the packet in unixtime.
	time_t* timestamp;
	/// NMEA checksum validity indicator, 1 = valid.
	short* nmea_checksum_ok;
	
	/// Weather report.
	fap_wx_report_t* wx_report;
	
	/// Telemetry report.
	fap_telemetry_t* telemetry;
	
	/// Message bits in case of mic-e packet.
	char* messagebits;
	/// Status message. No 0-termination.
	char* status;
	/// Amount of bytes in status message.
	unsigned int status_len;
	/// Capabilities list. Indexes 0, 2, 4, ... store keys and 1, 3, 5, ... the values (or NULL if the key has no value).
	char** capabilities;
	/// Amount of capabilities in list.
	unsigned int capabilities_len;
	
} fap_packet_t;




/// The parser.
/**
 * Resulting packet object will be filled with as much data as possible
 * based on the packet given as parameter. Unfilled fields are set to NULL.
 *
 * When parsing in AX.25 mode, source callsign and path elements are checked
 * to be strictly compatible with AX.25 specs so that they can be sent into
 * AX.25 network. Destination callsign is always checked this way.
 *
 * The parser should handle null bytes, newline chars and other badness
 * sometimes seen in packets without crashing as long as input_len is given
 * correctly.
 *
 * \param input TNC-2 mode APRS packet string.
 * \param input_len Amount of bytes in input.
 * \param is_ax25 If 1, packet is parsed as AX.25 network packet. If 0, packet is parsed as APRS-IS packet.
 * \return A packet is always returned, no matter how the parsing went. Use
 * error_code to check how it did. If library is not initialized, returns
 * NULL.
*/
fap_packet_t* fap_parseaprs(char const* input, unsigned int const input_len, short const is_ax25);


/// Return human-readable error message for given error code.
/**
 * \param error Error code from fap_packet_t.
 * \param buffer Pre-allocated space for the message. Must be at least 60 bytes long.
*/
void fap_explain_error(fap_error_code_t const error, char* buffer);

/// Convert mic-e message bits (three numbers 0-2) to a textual message.
/**
 * \param bits Bits as returned by fap_parseaprs().
 * \param buffer Pre-allocated space for the message. Must be at least 20 bytes long.
*/
void fap_mice_mbits_to_message(char const* bits, char* buffer);

/// Calculate distance between given locations.
/**
 * Returns the distance in kilometers between two locations given in decimal
 * degrees. Arguments are given in order as lon0, lat0, lon1, lat1, east and
 * north positive. The calculation uses the great circle distance, it is not
 * too exact, but good enough for us.
*/
double fap_distance(double lon0, double lat0, double lon1, double lat1);


/// Calculate direction from first to second location.
/**
 * Returns the initial great circle direction in degrees from lat0/lon0 to
 * lat1/lon1. Locations are input in decimal degrees, north and east
 * positive.
*/
double fap_direction(double lon0, double lat0, double lon1, double lat1);


/// Count amount of digihops the packet has gone through.
/**
 * The number returned is just an educated guess, not absolute truth.
 *
 * \return Number of digipeated hops or -1 in case of error.
*/
int fap_count_digihops(fap_packet_t const* packet);


/// Check if the callsign is a valid AX.25 callsign.
/**
 * \param input Callsign to be checked. If SSID is 0, the "-0" suffix can be omitted.
 * \param add_ssid0 If 1, a missing SSID 0 (in practice "-0") is appended to the
 * returned callsign. If 0, valid callsign is returned as is.
 * \return Given input if it was valid. NULL if the input was not a valid
 *  AX.25 address or library is not initialized.
 *
 * Please note that it's very common to use invalid callsigns on the
 * APRS-IS.
*/
char* fap_check_ax25_call(char const* input, short const add_ssid0);


/// Convert a KISS-frame into a TNC-2 compatible UI-frame.
/**
 * Non-UI and non-pid-F0 frames are dropped. The KISS-frame to be decoded
 * may or may not have a FEND (0xC0) character at beginning. If there's a
 * FEND in the frame before or at the end, the frame is cutted just before
 * the FEND. Byte unstuffing must not be done before calling this function.
 *
 * \param kissframe KISS-frame.
 * \param kissframe_len Amount of bytes in kissframe. Must be less than 512.
 * \param tnc2frame Result of conversion is stored here. In case of error an error message may be found here.
 * \param tnc2frame_len Amount of bytes available in tnc2frame is read from
 * here. Amount of bytes written to tnc2frame is written to here upon
 * return.
 * \param tnc_id TNC ID from the kissframe, usually zero.
 * \return 1 in case of success, 0 when error occured.
*/
int fap_kiss_to_tnc2(char const* kissframe, unsigned int kissframe_len,
                     char* tnc2frame, unsigned int* tnc2frame_len, unsigned int* tnc_id);


/// Convert a TNC-2 compatible UI-frame into a KISS data frame.
/**
 * The frame will be complete, i.e. it has byte stuffing done and FEND
 * (0xC0) characters on both ends.
 *
 * \param tnc2frame TNC-2 frame.
 * \param tnc2frame_len Amount of bytes in tnc2frame.
 * \param tnc_id TNC ID to use in KISS frame. When is doubt, use zero.
 * \param kissframe Where to store the KISS-frame. Be sure to allocate enough space.
 * \param kissframe_len Amount of bytes stored into kissframe during successfull conversion.
 * \return 1 in case of success, 0 when error occured.
*/
int fap_tnc2_to_kiss(char const* tnc2frame, unsigned int tnc2frame_len, unsigned int const tnc_id,
                     char* kissframe, unsigned int* kissframe_len);



/* Implementation-specific functions. */

/// Convert raw AX.25 frame into TNC-2 compatible UI-frame.
/**
 * Params and return value work just like similar ones in fap_kiss_to_tnc2().
*/
int fap_ax25_to_tnc2(char const* ax25frame, unsigned int ax25frame_len,
                     char* tnc2frame, unsigned int* tnc2frame_len);

/// Convert TNC-2 compatible UI-frame into raw AX.25-frame.
/**
 * Params and return value work just like similar ones in fap_tnc2_to_kiss().
*/
int fap_tnc2_to_ax25(char const* tnc2frame, unsigned int tnc2frame_len,
                     char* ax25frame, unsigned int* ax25frame_len);


/// Custom free() for fap_packet_t*.
/**
 * Use this in place of normal free() when disposing a packet. Will free()
 * all non-NULL fields for you.
*/
void fap_free(fap_packet_t* packet);



/// Library initialization.
/**
 * This must be called once and before anything else when starting to use
 * libfap in your program.
*/
void fap_init();



/// Library cleanup.
/**
 * This must be called once when closing your app. Make sure that no fap
 * calls are issued after this.
*/
void fap_cleanup();


#ifdef __cplusplus
}
#endif 

#endif // FAP_H


src/regs.h
/* $Id: regs.h 179 2011-01-17 12:38:00Z oh2gve $
 *
 * Copyright 2005, 2006, 2007, 2008, 2009, 2010 Tapio Sokura
 * Copyright 2007, 2008, 2009, 2010 Heikki Hannikainen
 *
 * Perl-to-C modifications
 * Copyright 2009, 2010, 2011 Tapio Aaltonen
 *
 * This file is part of libfap.
 *
 * Libfap is free software; you can redistribute it and/or modify it under the
 * terms of either:
 *
 * a) the GNU General Public License as published by the Free Software
 * Foundation; either version 1, or (at your option) any later
 * version, or
 * 
 * b) the "Artistic License". 
 * 
 * Both licenses can be found in the licenses directory of this source code
 * package.
 *
 * APRS is a registered trademark of APRS Software and Bob Bruninga, WB4APR.
*/

/**
 * \file regs.h
 * \brief Regular expressions needed in helpers.
 * \author Tapio Aaltonen
*/

#ifndef REGS_H
#define REGS_H

#include <regex.h>


extern regex_t fapint_regex_ax25call, fapint_regex_header, fapint_regex_digicall, fapint_regex_digicallv6;
extern regex_t fapint_regex_normalpos, fapint_regex_normalamb, fapint_regex_timestamp;
extern regex_t fapint_regex_mice_dstcall, fapint_regex_mice_body, fapint_regex_mice_amb;
extern regex_t fapint_regex_comment, fapint_regex_phgr, fapint_regex_phg, fapint_regex_rng, fapint_regex_altitude;
extern regex_t fapint_regex_mes_dst, fapint_regex_mes_ack, fapint_regex_mes_nack;
extern regex_t fapint_regex_wx1, fapint_regex_wx2, fapint_regex_wx3, fapint_regex_wx4, fapint_regex_wx5;
extern regex_t fapint_regex_wx_r1, fapint_regex_wx_r24, fapint_regex_wx_rami;
extern regex_t fapint_regex_wx_humi, fapint_regex_wx_pres, fapint_regex_wx_lumi, fapint_regex_wx_what;
extern regex_t fapint_regex_wx_snow, fapint_regex_wx_rrc, fapint_regex_wx_any, fapint_regex_wx_soft;
extern regex_t fapint_regex_nmea_chksum, fapint_regex_nmea_dst, fapint_regex_nmea_time, fapint_regex_nmea_date;
extern regex_t fapint_regex_nmea_specou, fapint_regex_nmea_fix, fapint_regex_nmea_altitude, fapint_regex_nmea_flag, fapint_regex_nmea_coord;
extern regex_t fapint_regex_telemetry, fapint_regex_peet_splitter, fapint_regex_kiss_callsign;


#endif // REGS_H


src/helpers.h
/* $Id: helpers.h 199 2011-05-22 05:09:13Z oh2gve $
 *
 * Copyright 2005, 2006, 2007, 2008, 2009, 2010 Tapio Sokura
 * Copyright 2007, 2008, 2009, 2010 Heikki Hannikainen
 *
 * Perl-to-C modifications
 * Copyright 2009, 2010, 2011 Tapio Aaltonen
 *
 * This file is part of libfap.
 *
 * Libfap is free software; you can redistribute it and/or modify it under the
 * terms of either:
 *
 * a) the GNU General Public License as published by the Free Software
 * Foundation; either version 1, or (at your option) any later
 * version, or
 * 
 * b) the "Artistic License". 
 * 
 * Both licenses can be found in the licenses directory of this source code
 * package.
 *
 * APRS is a registered trademark of APRS Software and Bob Bruninga, WB4APR.
*/

/**
 * \file helpers.h
 * \brief Declarations of helper functions for fap.c.
 * \author Tapio Aaltonen
*/

#ifndef HELPERS_H
#define HELPERS_H


#include <time.h>
#include "fap.h"



/// Parses packet->header into source, target and path.
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_header(fap_packet_t* packet, short const is_ax25);


/// Try to parse given packet body as mic-e packet.
/**
 * Everything after the mic-e report is also parsed here.
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_mice(fap_packet_t* packet, char const* input, unsigned int const input_len);


/// Returns unixtime based on timestamp in given body.
/**
 * \return Unixtime if body had valid timestamp, 0 if not.
*/
time_t fapint_parse_timestamp(char const* input);


/// Try to parse given packet body as compressed position report.
/**
 * Handles only compressed position report from beginning of body. Input is assumed to be null-terminated.
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_compressed(fap_packet_t* packet, char const* input);


/// Try to parse given packet body as normal position report.
/**
 * Handles only normal position report from beginning of body. Input is assumed to be null-terminated.
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_normal(fap_packet_t* packet, char const* input);


/// Check given packet body for course, speed PHGR, altitude and comments.
/**
 * Looks for optional fields and can't thus fail.
*/
void fapint_parse_comment(fap_packet_t* packet, char const* input, unsigned int const input_len);


/// Parse NMEA packet.
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_nmea(fap_packet_t* packet, char const* input, unsigned int const input_len);


/// Try to parse given packet body as object.
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_object(fap_packet_t* packet, char const* input, unsigned int const input_len);


/// Try to parse given packet body as item.
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_item(fap_packet_t* packet, char const* input, unsigned int const input_len);


/// Try to parse given packet body as message.
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_message(fap_packet_t* packet, char const* input, unsigned int const input_len);


/// Parse capabilities message.
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_capabilities(fap_packet_t* packet, char const* input, unsigned int const input_len);


/// Parse status message.
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_status(fap_packet_t* packet, char const* input, unsigned int const input_len);


/// Parse wx report.
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_wx(fap_packet_t* packet, char const* input, unsigned int const input_len);


/// Parse telemetry message.
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_telemetry(fap_packet_t* packet, char const* input);


/// Parses a Peet bros Ultimeter weather logging frame (!! header).
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_wx_peet_logging(fap_packet_t* packet, char const* input);


/// Parses a Peet bros Ultimeter weather packet ($ULTW header).
/**
 * \return 1 on success, 0 on failure.
*/
int fapint_parse_wx_peet_packet(fap_packet_t* packet, char const* input);


/// Parse given !DAO! extension.
/**
 * Parse a possible !DAO! extension (datum and extra lat/lon digits).
 * \param packet Results are saved here.
 * \param input The 3 bytes between exclamation marks.
 * \return 1 if a valid !DAO! extension was detected in the test subject (and stored in $rethash), 0 if not.
*/
int fapint_parse_dao(fap_packet_t* packet, char input[3]);



/// Validates given KISS-level callsign.
/**
 * Checks a callsign for validity and strips trailing spaces out and returns
 * the string.
 * \return Copy of input at success, NULL when failure occurs.
*/
char* fapint_check_kiss_callsign(char* input);



/* Implementation-specific helpers. */


/// Creates and initializes an empty packet.
fap_packet_t* fapint_create_packet();



#endif // HELPERS_H


src/helpers2.h
/* $Id: helpers2.h 199 2011-05-22 05:09:13Z oh2gve $
 *
 * Copyright 2005, 2006, 2007, 2008, 2009, 2010 Tapio Sokura
 * Copyright 2007, 2008, 2009, 2010 Heikki Hannikainen
 *
 * Perl-to-C modifications
 * Copyright 2009, 2010 Tapio Aaltonen
 *
 * This file is part of libfap.
 *
 * Libfap may be copied only under the terms of either the Artistic License
 * or the GNU General Public License, which may be found in the libfap
 * source kit.
 *
 * APRS is a registered trademark of APRS Software and Bob Bruninga, WB4APR.
*/

/**
 * \file helpers2.h
 * \brief Declarations of helper functions for helpers.c.
 * \author Tapio Aaltonen
*/

#ifndef HELPERS2_H
#define HELPERS2_H


#include "fap.h"


/// Max amount of path elements.
#define MAX_DIGIS 8

/// Nautical miles per hour to kilometers per hour.
#define KNOT_TO_KMH 1.852

/// Miles per hour to kilometers per hour.
#define MPH_TO_KMH 1.609344

/// Kilometers per hour to meters per second.
#define KMH_TO_MS 10.0/36.0

/// Miles per hout to meters per second.
#define MPH_TO_MS (MPH_TO_KMH * KMH_TO_MS)

/// Feets to meters.
#define FT_TO_M 0.3048

/// Fahrenheit to celcius degrees.
#define FAHRENHEIT_TO_CELCIUS(x) ((x-32.0)/1.8)

/// Hundredths of an inch to millimeters
#define HINCH_TO_MM 0.254



/// Linked list item.
typedef struct fapint_llist_item
{
	/// Data stored in the element.
	char* text;
	/// Next element in list.
	struct fapint_llist_item* next;
} fapint_llist_item_t;


/// Cleanup function for linked list.
void fapint_clear_llist(fapint_llist_item_t* list);



/// Return position resolution in meters based on the number of minute decimal digits.
/** Also accepts negative numbers, i.e. -1 for 10 minute resolution and -2
 *  for 1 degree resolution. Calculation is based on latitude so it is worst
 *  case (resolution in longitude gets better as you get closer to the
 *  poles).
*/
double fapint_get_pos_resolution(int const minute_digit_count);


/// Parse station symbol from GPSxyz or SPCxyz -formatted destination callsign.
int fapint_parse_symbol_from_dst_callsign(fap_packet_t* packet);


/// Maps destination callsign symbol chars to APRS symbol.
/**
 * Contains a list of mappings from GPSxyz (or SPCxyz) to APRS symbols.
 * Overlay characters (z) are not handled here.
 * \param input Two chars from destination callsign.
 * \param output Two chars ([0] and [1]) of aprs symbol.
*/
int fapint_symbol_from_dst_type(char input[2], char* output);


/// Returns 0 if string contains non-numeric chars, 1 if not.
int fapint_is_number(char const* input);


/// Returns 1 if given date is valid. */
int fapint_check_date(unsigned int year, unsigned int month, unsigned int day);


/// Get coordinate from nmea fields.
/**
 * Field 2 is used to decide between latitude and longitude. If
 * pos_resolution is not set, this will set it too.
 * \param packet Results are saved here.
 * \param field1 The (dd)dmm.m(mmm..) string.
 * \param field2 The north/south or east/west indicator.
 * \return Zero for failure, non-zero for success.
*/
int fapint_get_nmea_latlon(fap_packet_t* packet, char* field1, char* field2);


/// Parse base-91 telemetry from comment text.
/**
 * \param packet Results are saved here.
 * \param rest Comment text.
 * \param rest_len Lenght of rest in bytes.
*/
void fapint_parse_comment_telemetry(fap_packet_t* packet, char** rest, unsigned int* rest_len);


/// Set all fields of given wx report to NULL.
void fapint_init_wx_report(fap_wx_report_t* wx_report);


// Reset fields of given telemetry report.
/**
 * Numeric fields are set to 0, bits are set to ?.
 * \param report to reset.
*/
void fapint_init_telemetry_report(fap_telemetry_t* telemetry);


/// Returns copy of given input with specified part removed.
char* fapint_remove_part(char const* input, unsigned int const input_len,
                         unsigned int const part_so, unsigned int const part_eo,
                         unsigned int* result_len);


#endif // HELPERS2_H
